### PR TITLE
perf(Backup): Improve backup performance

### PR DESCRIPTION
**Pull request by jarifibrahim**
This PR has the following changes
- Use thread-local buffer for codecs
- Add new ReadTs field to Manifest and use it instead of sinceTs
- Use snappy for backups

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7601)

### CODE REVIEW

File: posting/list.go [hunk 1]
```diff
@@ -868,15 +868,12 @@ func (l *List) ToBackupPostingList(bl *pb.BackupPostingList, alloc *z.Allocator)
...
-		bl.UidBytes = buf.Bytes()
-	} else {
-		bl.Uids = codec.Decode(ol.Pack, 0)
-	}
+
+	// Encode uids to []byte instead of []uint64. This helps improve memory usage.
+	buf.Reset()
+	codec.DecodeToBuffer(buf, ol.Pack)
+	bl.UidBytes = buf.Bytes()
+
```
Comment (NamanJain8) [inline posting/list.go:876]: Why did we change this heuristic? This change would mean that `bl.Uid` remains for backward compatibility only.

File: worker/file_handler.go [hunk 2]
```diff
@@ -198,15 +199,20 @@ func (h *fileHandler) Load(uri *url.URL, backupId string, backupNum uint64, fn l
...
+					compression:    manifest.Compression,
+				})
 			if err != nil {
 				return LoadResult{Err: err}
 			}
 			maxUid = x.Max(maxUid, groupMaxUid)
 			maxNsId = x.Max(maxNsId, groupMaxNsId)
 		}
-		since = manifest.Since
+		since = manifest.ReadTs
```
Comment (NamanJain8) [inline worker/file_handler.go:215]: `manifest.ValidReadTs()`

File: worker/backup_common.go [hunk 3]
```diff
@@ -71,6 +71,15 @@ type Manifest struct {
 	// DropOperations lists the various DROP operations that took place since the last backup.
 	// These are used during restore to redo those operations before applying the backup.
 	DropOperations []*pb.DropOperation `json:"drop_operations"`
+	// Compression keeps track of the compression that was used for the data.
+	Compression string `json:"compression"`
+}
+
+func (m *Manifest) ValidReadTs() uint64 {
```
Comment (NamanJain8) [inline worker/backup_common.go:78]: Add function description.

**Review by NamanJain8 — COMMENTED** [review]
Got few comments. The benefit of introducing `ReadTs` in the manifest isn't clear to me. Can you please add that in the PR description?

File: worker/file_handler.go (hunk 2 above)
Comment (jarifibrahim) [reply worker/file_handler.go]: Done in 40558331a

File: worker/backup_common.go (hunk 3 above)
Comment (jarifibrahim) [reply worker/backup_common.go:81]: Done in 40558331a

**Review by jarifibrahim — COMMENTED** [review]
@NamanJain8 The readTs is added for better readability. The sinceTs used in actually the readTs in case of backups.

**Review by NamanJain8 — APPROVED** [review]

**Comment by NamanJain8** [comment]
Looks good to me.

**Pull request merged**